### ISSUE TITLE
Tkinter - Can't find a usable init.tcl

### ISSUE THREAD

**Issue by oldium** [python/cpython#111754]
# Bug report

### Bug description:

I tried to add `widgetPlus` from [Tklib](https://www.tcl.tk/software/tklib/) and its dependency `snit` from [Tcllib](https://www.tcl.tk/software/tcllib/) to my project, but executing `root.tk.call("package", "require", "widgetPlus")` after instantiating `root = tkinter.Tk()` failed with Tcl error:

```
_tkinter.TclError: Can't find a usable init.tcl in the following directories:
    C:/Users/oldium/AppData/Local/Programs/Python/Python312/lib/tcl8.6 C:/Users/oldium/AppData/Local/Programs/Python/lib/tcl8.6 C:/Users/oldium/AppData/Local/Programs/lib/tcl8.6 C:/Users/oldium/AppData/Local/Programs/Python/library C:/Users/oldium/AppData/Local/Programs/library C:/Users/oldium/AppData/Local/Programs/tcl8.6.13/library C:/Users/oldium/AppData/Local/tcl8.6.13/library
```

The real TCL installation path is `C:/Users/oldium/AppData/Local/Programs/Python/Python312/tcl/tcl8.6`, note `tcl` instead of `lib`.

Strangely, the same application worked when started as `exe` after [PyInstaller](https://pypi.org/project/pyinstaller/) was used. I tracked the problem down to the minimal test case (as used by Tcllib module `snit`):

```python
import glob
import os
import sys
import tkinter as tk
from pathlib import Path

if __name__ == "__main__":
    tcl = tk.Tcl()
    interpreter = tcl.call("interp", "create")
    tcl.call("interp", "delete", interpreter)
    tcl.call("puts", "It worked!")
```

I also found a workaround (I used `glob` to prevent fixing on a particular Tcl/Tk version), which works both in “normal” and `venv` environments:

```
# Fix Tcl/Tk folder
os.environ["TCL_LIBRARY"] = str(Path(glob.glob(os.path.join(sys.base_prefix, "tcl", "tcl*", "init.tcl"))[0]).parent)
os.environ["TK_LIBRARY"] = str(Path(glob.glob(os.path.join(sys.base_prefix, "tcl", "tk*", "pkgIndex.tcl"))[0]).parent)
os.environ["TIX_LIBRARY"] = str(Path(glob.glob(os.path.join(sys.base_prefix, "tcl", "tix*", "pkgIndex.tcl"))[0]).parent)
```

It looks like the Tcl/Tk installation works fine but needs additional configuration on the Python side. And please note that the workaround is usable for the Python's Tcl/Tk installation and does not consider existing Tcl/Tk installations and existing `*_LIBRARY` environmental variables.

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

Windows

**Comment by oldium** [comment]
This works in PyInstaller's generated `exe` because it sets the TCL_LIBRARY and TK_LIBRARY by itself as seen in Process Explorer:

![image](https://github.com/python/cpython/assets/1260936/5cf75e33-a7a8-4744-adbc-af6e38f405a0)

Maybe Python could do the same as PyInstaller does, i.e. set `TCL_LIBRARY` and `TK_LIBRARY` (and `TIX_LIBRARY`?) if they are not already set?

**Comment by FujiwaraChoki** [comment]
`brew install tcl-tk` fixed it for me.

**Comment by oldium** [comment]
> `brew install tcl-tk` fixed it for me.

On Windows? 🤔

**Comment by FujiwaraChoki** [comment]
> > `brew install tcl-tk` fixed it for me.
> 
> On Windows? 🤔
> 
> Tcl/Tk is part of Python installer on Windows already.
> 
> Edit: You are talking about separate installation of Tcl/Tk, so you probably already have the environment variables set correctly, but I am talking about embedded version of Tcl/Tk installed by the Python installer on Windows, which resides in the Python installation directory.

Yeah I'm on an M3 Mac, so not Windows. The issue is likely that there is a discrepancy between the Python and TCL/TK Versions. You need to re-install (re-build) Python everytime you install another version of TCL/TK.

I do not know the solution for windows.

**Comment by oldium** [comment]
> > > `brew install tcl-tk` fixed it for me.
> > 
> > 
> > On Windows? 🤔
> > Tcl/Tk is part of Python installer on Windows already.
> > Edit: You are talking about separate installation of Tcl/Tk, so you probably already have the environment variables set correctly, but I am talking about embedded version of Tcl/Tk installed by the Python installer on Windows, which resides in the Python installation directory.
> 
> Yeah I'm on an M3 Mac, so not Windows. The issue is likely that there is a discrepancy between the Python and TCL/TK Versions. You need to re-install (re-build) Python everytime you install another version of TCL/TK.

There is no discrepancy in versions. The Tcl/Tk runtime simply cannot find the installation library, because it expects it to be in `<Python Dir>/lib/tcl8.6` (some auto-detection), but it is actually stored in `<Python Dir>/tcl/tcl8.6`. So the Tcl autodiscovery does not find it.

> I do not know the solution for windows.

I already mentioned possible solution/workaround in the issue description - set the environment variables TCL_LIBRARY/TK_LIBRARY/TIX_LIBRARY by Python if they are not already set.

**Comment by FujiwaraChoki** [comment]
> > > > `brew install tcl-tk` fixed it for me.
> > > 
> > > 
> > > On Windows? 🤔
> > > Tcl/Tk is part of Python installer on Windows already.
> > > Edit: You are talking about separate installation of Tcl/Tk, so you probably already have the environment variables set correctly, but I am talking about embedded version of Tcl/Tk installed by the Python installer on Windows, which resides in the Python installation directory.
> > 
> > 
> > Yeah I'm on an M3 Mac, so not Windows. The issue is likely that there is a discrepancy between the Python and TCL/TK Versions. You need to re-install (re-build) Python everytime you install another version of TCL/TK.
> 
> There is no discrepancy in versions. The Tcl/Tk runtime simply cannot find the installation path, because it expects it to be in `<Python Dir>/lib/tcl8.6` (some auto-detection), but it is actually stored in `<Python Dir>/tcl/tcl8.6`. So the Tcl autodiscovery does not find it.
> 
> > I do not know the solution for windows.
> 
> I already mentioned possible solution/workaround in the issue description - set the environment variables `TCL_LIBRARY`/`TK_LIBRARY`/`TIX_LIBRARY` by Python if they are not already set.

I know it's because of the paths.. The path had the wrong version in it which is why it wasn't working for me. Also I'm not on windows.

**Comment by oldium** [comment]
> > > > > `brew install tcl-tk` fixed it for me.
> > > > 
> > > > 
> > > > On Windows? 🤔
> > > > Tcl/Tk is part of Python installer on Windows already.
> > > > Edit: You are talking about separate installation of Tcl/Tk, so you probably already have the environment variables set correctly, but I am talking about embedded version of Tcl/Tk installed by the Python installer on Windows, which resides in the Python installation directory.
> > > 
> > > 
> > > Yeah I'm on an M3 Mac, so not Windows. The issue is likely that there is a discrepancy between the Python and TCL/TK Versions. You need to re-install (re-build) Python everytime you install another version of TCL/TK.
> > 
> > 
> > There is no discrepancy in versions. The Tcl/Tk runtime simply cannot find the installation path, because it expects it to be in `<Python Dir>/lib/tcl8.6` (some auto-detection), but it is actually stored in `<Python Dir>/tcl/tcl8.6`. So the Tcl autodiscovery does not find it.
> > > I do not know the solution for windows.
> > 
> > 
> > I already mentioned possible solution/workaround in the issue description - set the environment variables `TCL_LIBRARY`/`TK_LIBRARY`/`TIX_LIBRARY` by Python if they are not already set.
> 
> I know it's because of the paths.. The path had the wrong version in it which is why it wasn't working for me. Also I'm not on windows.

Correct, your comments are off-topic. You have a different problem on a different operating system. 🤷
